index.js

```diff
@@ -548,20 +637,50 @@ class KeyringController extends EventEmitter {
    * @param {string} password - The keyring controller password.
    * @returns {Promise<Array<Keyring>>} The keyrings.
    */
-  async unlockKeyrings(password) {
+  async unlockKeyrings(password, encryptionKey) {
     const encryptedVault = this.store.getState().vault;
     if (!encryptedVault) {
       throw new Error('Cannot unlock without a previous vault.');
     }
 
+    if (password === undefined && encryptionKey === undefined) {
+      throw new Error(
+        'No way to decrypt a salted vault without a password or encrypted key',
+      );
+    }
+
     await this.clearKeyrings();
-    const vault = await this.encryptor.decrypt(password, encryptedVault);
-    this.password = password;
+
+    // MV3: If the separator string is in the vault string, the user has already migrated
+    // from the previous password-only model
+    const vault = await this.attemptGetDecryptedVault(
+      encryptedVault,
+      password,
+      encryptionKey,
+    );
+
     await Promise.all(vault.map(this._restoreKeyring.bind(this)));
+
     await this._updateMemStoreKeyrings();
+
     return this.keyrings;
   }
 
+  parseVault(encryptedVault) {
+    const [vault, salt] = encryptedVault.split(VAULT_SEPARATOR);
+    return { vault, salt };
+  }
+
+  // MV3:  Generates the encrypted key
+  async _generateEncryptionKey(password, salt) {
+    return toHex(sha256(utf8ToBytes(password + salt)));
```

Comment:
Why are we using `ethereum-cryptography/sha256` here instead of the encryptor?

Reply:
@darkwing I know this code came from the branch I pushed up, but you might have missed this comment in our DM with Zb:

> I think the shim could be used for stubbing random salt with predictable values and asserting for it.
Also, getting a dependency instead using a feature in the browser is not only a detriment to maintainability but also performance :shushing_face:

So we should go back to your old implementation that uses the native crypto api directly. But as per Mark's comment, we can probably just use the encryptor itself

Reply:
to be more specific, you could use `keyFromPassword(password, salt)`

Reply:
But, we might want to consider how we can more fully leverage browser passworder and push more logic down to that

Reply:
I tried doing the following:

```
async _generateEncryptionKey(password, salt) {
    //return toHex(sha256(utf8ToBytes(password + salt)));
    const key = await encryptor.keyFromPassword(password, salt);
    const exportedKey = await window.crypto.subtle.exportKey('jwk', key);
    const keyString = new Uint8Array(exportedKey).toString();

    console.log("key is: ", key, exportedKey);
    console.log("keystring is: ", keyString);
    return key;
  }
```

Result being:

<img width="804" alt="AttemptedKeyExport" src="https://user-images.githubusercontent.com/46655/192866720-b8f88322-4ef0-4023-94ee-79c10876baa1.png">

The string form of extracted key is empty.  I don't think this helps us.

Reply:
It's not totally clear what you're trying to do in that snippet, or how it relates to this thread. But that is failing because you're asking for the key to be exported as a JWT, which is an object. Then you're passing the object into the Uint8Array constructor, which isn't meant to accept that sort of object as input. It tries to iterate over it but it's not iterable, giving you an empty arraybuffer, thus the empty string.